_posts/2020-10-24-text-layout.md

```diff
@@ -0,0 +1,217 @@
+---
+layout: post
+title:  "Text layout is a loose hierarchy of segmentation"
+date:   2020-10-25 15:00:42 -0700
+categories: [text]
+---
+I love text layout, and have been working with it in one form or other for over 35 years. Yet, knowledge about it is quite arcane. I don't believe there is a single place where it's all properly written down. I have some explanation for that: while basic text layout is very important for UI, games, and other contexts, a lot of the "professional" needs around text layout are embedded in *much* more complicated systems such as Microsoft Word or a modern Web browser.
+
+A complete account of text layout would be at least a small book. Since there's no way I can write that now, this blog post is a small step towards that - in particular, an attempt to describe the "big picture," using the conceptual framework of a "loose hierarchy." Essentially, a text layout engine breaks the input into finer and finer grains, then reassembles the results into a text layout object suitable for drawing, measurement, and hit testing.
+
+The main hierarchy is concerned with laying out a single strip of text. Line breaking is also important, but has a separate, parallel hierarchy.
+
+## The main text layout hierarchy
+
+The hierarchy is:
+- (a) paragraph segmentation
+- (b) rich text style
+- (c) BiDi
+- (d) itemization (coverage by font)
+- (e) script
+- (f) shaping (cluster)
```

Comment:
Also on the markup employed: I’m a markup purist, and strongly dislike the semantics of using an unordered list with ordered markers as text for this, and the visual result of “• (a)”. The proper technique is markup like this:

```html
<style>
    @counter-style parenthesised-lower-alpha {
        system: extends lower-alpha;
        prefix: "(";
        suffix: ")  ";
    }

    ol[type="a"] {
        list-style-type: lower-alpha; /* “a. ” (fallback for browsers that don’t support @counter-style) */
        list-style-type: parenthesised-lower-alpha; /* “(a)  ” */
    }
</style>

<ol type="a">
    <li>…
    <li>…
    <li>…
</ol>
```

(Markdown is bad enough that you’d need to drop into HTML here for the list. For comparison, reStructuredText pays attention to whether you wrote “1.”, “a)” or “(i)” in the source, and its HTML writer turns that into the right ol[type], even if it sadly doesn’t try to make a @counter-style for each prefix/system/suffix combo.)

But unfortunately, although Firefox implemented CSS Counter Styles 3 six years ago, no one else has followed suit. So the whole world is *still* reaching for awful `::before` and `counter()` hacks whenever they want to do things like this, if they don’t like a fallback of “a.” or “1.” in other browsers.

This is really just a bit of a rant, decide for yourself whether you want to do anything with it! 🙂

Reply:
I got rid of the list, it was only a stand-in for the diagram.